pin-project-internal/src/pin_project/args.rs

```diff
@@ -190,10 +205,28 @@ impl Parse for Args {
                 ));
             }
         }
-        let project_replace = match (project_replace_span, project_replace_value) {
-            (None, _) => ProjReplace::None,
-            (Some(span), Some(ident)) => ProjReplace::Named { ident, span },
-            (Some(span), None) => ProjReplace::Unnamed { span },
+        let project_replace =
+            match (project_replace_span, project_replace_value, project_replace_pub) {
+                (None, _, _) => ProjArgs::None,
+                (Some(span), Some(ident), false) => ProjArgs::Named { ident, span },
+                (Some(span), Some(ident), true) => ProjArgs::NamedPublic { span, ident },
+                (Some(span), None, false) => ProjArgs::Unnamed { span },
+                (Some(span), None, true) => {
+                    return Err(Error::new(
+                        span,
+                        "project_replace cannot be pub if it is not named",
+                    ));
```

Comment:
If we reach this point, it's a bug in our code, not the user's fault, so `unreachable!` would be better here.